libs/grout/src/server.ts

```diff
@@ -0,0 +1,176 @@
+ */
+export class GroutError extends Error {}
+
+/**
+ * Registers route handlers for each RPC method in a Grout API on an Express
+ * router. This allows you to easily mount the Grout API within a larger
+ * Express app like so:
+ *
+ *  const api = createApi({ ... methods ... });
+ *  const app = express();
+ *  router = express.Router();
+ *
+ *  // The router must use the text body parser middleware with content type
+ *  // 'application/json'. Make sure you don't add any other body parsers
+ *  // upstream of the Grout router (e.g. with app.use)
+ *  router.use(express.text({ type: 'application/json' }));
+ *
+ *  registerRoutes(api, router);
+ *
+ *  // Here, we mount the Grout API at /api, but you can mount it anywhere.
+ *  app.use('/api', groutRouter);
+ *
+ * All routes will use the POST HTTP method with a JSON body (for RPC input) and
+ * return a JSON response (for RPC output), using Grout's serialization format
+ * (see the serialization module for details). This allows requests to be
+ * human-readable while still supporting richer types than plain JSON.
+ *
+ * Unexpected exceptions will return a 500 status code with a plain JSON
+ * response containing the error mesage. This will cause the client to throw an
+ * exception.
+ *
+ * You should NOT throw exceptions from your RPC methods for known errors (e.g.
+ * invalid input). Instead, use the Result type to return known errors.  This
+ * ensures that consumers of the method will be forced to handle that error case
+ * explicitly.
+ */
+export function registerRoutes(api: AnyApi, router: express.Router): void {
+  if (!router.stack.some((layer) => layer.name === 'textParser')) {
+    throw new GroutError(
+      'Grout requires the router to use the text body parser middleware with content type application/json.' +
+        " Add it like so: app.use(express.text({ type: 'application/json' }))."
+    );
+  }
+
+  for (const [methodName, method] of Object.entries<AnyRpcMethod>(api)) {
+    const path = `/${methodName}`;
+    debug(`Registering route: ${path}`);
+
+    // All routes use the POST method. This doesn't quite follow the traditional
+    // semantics of HTTP, since there may or may not be a side-effect. But it's
+    // better to err on the side of possible side-effects (as opposed to use GET).
+    // We don't try to use the correct HTTP method because that would require
+    // some way to annotate RPC methods with the appropriate method, which is
+    // more complexity for little practical gain.
+    router.post(path, async (request, response) => {
+      try {
+        debug(`Call: ${methodName}(${request.body})`);
+        if (!isString(request.body)) {
+          throw new GroutError(
+            'Request body was parsed as something other than a string.' +
+              ' Make sure you havent added any other body parsers upstream' +
+              ' of the Grout router - e.g. app.use(express.json()).' +
+              ` Body: ${JSON.stringify(request.body)}`
```

Comment:
Will this constraint make it hard to migrate apps to Grout gradually?

Reply:
In my initial migration of VxScan, I'm just moving all the other routes to a separate router as a first step so there's no body-parsing middleware on the top-level app. It's a pretty easy change to make.

Reply:
Ah nice, that makes sense!